### PR TITLE
Allow masking of internal errors

### PR DESCRIPTION
**Pull request by rolandshoemaker**
For deployments where there isn't something deployed in front of `ct_server` that filters out 500 error bodies you may not want to actually leak what the golang error was to the user (also it is 99% probably not useful to them) so allow just logging the error and returning `Internal Server Error` to the user.

### CODE REVIEW

File: trillian/ctfe/handlers.go [hunk 1]
```diff
@@ -178,14 +180,21 @@ func (a AppHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	// Additional check, for consistency the handler must return an error for non-200 status
 	if status != http.StatusOK {
-		glog.Warningf("%s: %s handler non 200 without error: %d %v", a.Context.LogPrefix, a.Name, status, err)
-		sendHTTPError(w, http.StatusInternalServerError, fmt.Errorf("http handler misbehaved, status: %d", status))
+		glog.Warningf("%s: %s handler non 200 without error: %d", a.Context.LogPrefix, a.Name, status)
+		err = fmt.Errorf("http handler misbehaved, status: %d", status)
+		if a.Context.instanceOpts.MaskInternalErrors {
+			err = errInternalServerError
+		}
+		sendHTTPError(w, http.StatusInternalServerError, err)
```
Comment (daviddrysdale) [inline trillian/ctfe/handlers.go]: If we made `sendHTTPError` a method on `*LogContext` I think the error replacement could be commonized in one place.

File: trillian/ctfe/handlers.go [hunk 2]
```diff
@@ -109,6 +109,8 @@ var (
 	rspLatency       monitoring.Histogram // logid, ep, rc => value
 )
 
+var errInternalServerError = errors.New("Internal Server Error")
```
Comment (daviddrysdale) [inline trillian/ctfe/handlers.go]: I think we need something blander/more vague for the error -- if a client submits something invalid (e.g. first > second) we don't want an error message that makes it sound like it's our fault.  Maybe something like "An error occurred processing this request" ?

**Comment by daviddrysdale** [comment]
This would fix #81

**Comment by Martin2112** [comment]
I thought we'd done this in a previous change but possibly not everywhere.

**Comment by rolandshoemaker** [comment]
Sorry for dropping this on the floor. My original intent with this was to only mask errors that would've returned `500 Internal Server Errror` but I seem to have accidentally done this for everything except `200`s. My main motivation is that we don't really want to leak RPC transport/backend errors that might contain non-public information.

If this makes sense I'll dive back in and make the requested changes and move it to only do it for 500s.

**Comment by Martin2112** [comment]
5xx is more likely to have internal details so that sounds reasonable to me.

Ideally 4xx errors should produce something useful for clients. If there are cases where this doesn't happen we should fix them but I think work was already done on that.

**Comment by RJPercival** [comment]
Do you think you'll have time to return to this, @rolandshoemaker?

**Comment by rolandshoemaker** [comment]
Oh boy, I completely forgot I had this up, sorry about that. I'll take a crack at getting this back up to date this weekend.

**Comment by rolandshoemaker** [comment]
Should be back in a reivew-able form. Took @daviddrysdale's suggestion to move the `sendHTTPError` method onto `logInfo` (previously `LogContext`) in order to centralize the masking logic.

File: trillian/ctfe/instance_test.go [hunk 3]
```diff
@@ -251,3 +254,20 @@ func TestSetUpInstanceSetsValidationOpts(t *testing.T) {
...
 	}
 }
+
+func TestErrorMasking(t *testing.T) {
+	info := logInfo{}
+	w := httptest.NewRecorder()
+	prefix := "Internal Server Error"
+	err := errors.New("well that's bad")
+	info.SendHTTPError(w, 500, err)
+	if w.Body.String() != fmt.Sprintf("%s\n%v\n", prefix, err) {
```
Comment (daviddrysdale) [inline trillian/ctfe/instance_test.go]: nit: probably worth using the `got, want := ...` thing here (and below):
```golang
	if got, want:= w.Body.String(), fmt.Sprintf("%s\n%v\n", prefix, err); got != want {
		t.Errorf("SendHTTPError: got %s, want %s", got, want)
￼ 	}
```

**Comment by rolandshoemaker** [comment]
Resolved comments, thanks.

**Comment by daviddrysdale** [comment]
Could you merge up to the latest `master` please?

**Pull request merged**